File: leetcode/AddTwoNumbers.java
public class AddTwoNumbers {

	public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {

		if (l1 == null) {
			return l2;
		} else if (l2 == null) {
			return l1;
		}

		int nextval = l1.val + l2.val;
		int carry = 0;
		if (nextval >= 10) {
			nextval -= 10;
			carry = 1;
		}

		ListNode head = new ListNode(nextval);
		ListNode next = head;

		while (l1.next != null && l2.next != null) {
			l1 = l1.next; l2 = l2.next;
			nextval = l1.val + l2.val + carry;
			if (nextval >= 10) {
				nextval -= 10;
				carry = 1;
			} else {
				carry = 0;
			}
			next.next = new ListNode(nextval);
			next = next.next;
		}

		if (l1.next != null) {
			while (l1.next != null) {
				l1=l1.next;
				nextval = l1.val + carry;
				if (nextval >= 10) {
					nextval -= 10;
					carry = 1;
				} else {
					carry = 0;
				}
				next.next = new ListNode(nextval);
				next = next.next;
			}
		} else {
			while (l2.next != null) {
				l2=l2.next;
				nextval = l2.val + carry;
				if (nextval >= 10) {
					nextval -= 10;
					carry = 1;
				} else {
					carry = 0;
				}
				next.next = new ListNode(nextval);
				next = next.next;
			}
		}

		if (carry == 1) {
			next.next = new ListNode(carry);
		}
		return head;

	}

	public static void main(String[] args) {
		ListNode a = new ListNode(new int[] {5});
		ListNode b = new ListNode(new int[] {5});
		addTwoNumbers(a,b).print();
	}


}

File: leetcode/Medium/addTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *sent = new ListNode(0), *t = sent;
        int co = 0;
        while (l1 && l2) {
            int s = l1->val + l2->val + co;
            if (s < 10) {
                t->next = new ListNode(s);
                co = 0; 
            } else {
                t->next = new ListNode(s-10);
                co = 1;
            }
            l1 = l1->next;
            l2 = l2->next;
            t = t->next;
        }
        while (l1) {
            int s = l1->val + co;
            if (s < 10) {
                t->next = new ListNode(s);
                co = 0; 
            } else {
                t->next = new ListNode(s-10);
                co = 1;
            }
            l1 = l1->next;
            t = t->next;
        }
        while (l2) {
            int s = l2->val + co;
            if (s < 10) {
                t->next = new ListNode(s);
                co = 0; 
            } else {
                t->next = new ListNode(s-10);
                co = 1;
            }
            l2 = l2->next;
            t = t->next;
        }
        if (co) {
            t->next = new ListNode(co);
        }
        return sent->next;
    }
};